[AutoDiff] Revamp array literal initialization differentiation.

dan-zheng opened the pull request:
Array literal initialization requires special reverse-mode differentiation
support because it is not a regular function application in SIL. Instead,
it is lowered as an `array.uninitialized_intrinsic` application and
`store`/`copy_addr` instructions that set individual array elements:

```
%result = apply %array.uninitialized_intrinsic : $(Array<T>, RawPointer)
(%array, %ptr) = destructure_tuple %result
%elt0 = pointer_to_address %ptr to $*T       // element address
%index_1 = integer_literal $Builtin.Word, 1
%elt1 = index_addr %elt0, %index_1           // element address
...
```

Previously, array literal initialization differentiation was handled as a
special case in `PullbackEmitter::visitApplyInst`.
`PullbackEmitter::visitArrayInitialization` visited active intrinsic
applications, found all `store` and `copy_addr` instructions writing to
array element addresses, and accumulated destination adjoints into
source adjoints.

This worked in cases when element values were computed before intrinsic
applications, but not when element values were computed after intrinsic
applications but before `store`/`copy_addr` instructions.

Now, array literal initialization differentiation is handled as
a special case in `PullbackEmitter` for `store` and `copy_addr`
instructions. For each special `store` and `copy_addr` instruction,
array element adjoints are accumulated into source adjoints.

Activity analysis now marks element result addresses (`pointer_to_address`
and `index_addr` instructions) of `array.uninitialized_intrinsic`
applications as active.

Add array literal initialization tests.

Add negative tests for the following issues:
- TF-975: multiple array literals.
- TF-977: array literals with `tuple_element_addr` elements.
- TF-978: apply with array literal address as indirect result.

Resolves TF-952.

Future improvements:
- TF-976: move "get adjoint buffer of array literal element" logic out of `PullbackEmitter::accumulateArrayElementAdjoint` into `PullbackEmitter::getAdjointProjection`.

---

Example:

```swift
struct Struct: Differentiable {
  var x, y: Float
  func method() -> [Float] {
    return [x, y]
  }
}
let s = Struct(x: 0, y: 0)
let v = Array<Float>.TangentVector([10, 20])
print(pullback(at: s, in: { s in s.method() })(v))
// Before: TangentVector(x: 0.0, y: 0.0)
// After:  TangentVector(x: 10.0, y: 20.0)
```

dan-zheng committed:
[AutoDiff] Revamp array literal initialization differentiation.

Array literal initialization requires special reverse-mode differentiation
support because it is not a regular function application in SIL. Instead,
it is lowered as an `array.uninitialized_intrinsic` application with by
`store`/`copy_addr` instructions to set individual array elements:

```
%result = apply %array.uninitialized_intrinsic : $(Array<T>, RawPointer)
(%array, %ptr) = destructure_tuple %result
%elt0 = pointer_to_address %ptr to $*T       // element address
%index_1 = integer_literal $Builtin.Word, 1
%elt1 = index_addr %elt0, %index_1           // element address
...
```

Previously, array literal initialization differentiation was handled as a
special case in `PullbackEmitter::visitApplyInst`.
`PullbackEmitter::visitArrayInitialization` visited active intrinsic
applications, found all `store` and `copy_addr` instructions writing to
array element addresses, and accumulated destination adjoints into
source adjoints.

This worked in cases when element values were computed before intrinsic
applications, but not when element values were computed after intrinsic
applications but before `store`/`copy_addr` instructions.

Now, array literal initialization differentiation is handled as
a special case in `PullbackEmitter` for `store` and `copy_addr`
instructions. For each special `store` and `copy_addr` instruction,
array element adjoints are accumulated into source adjoints.

`LinearMapInfo::shouldDifferentiateInstruction` now also has special
support for `store`-like instructions whose destination is an element
address of an `array.uninitialized_intrinsic` application: it returns
true if the intrinsic application (representing the semantic destination)
is active.

Add array literal initialization tests.

Resolves TF-952.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -92,9 +92,36 @@ static bool isArrayLiteralIntrinsic(ApplyInst *ai) {
-  if (auto *applyInst = dyn_cast<ApplyInst>(v))
-    if (isArrayLiteralIntrinsic(applyInst))
-      return applyInst;
+  if (auto *ai = dyn_cast<ApplyInst>(v))
```
I suggest braces around multiline bodies for clarity.

dan-zheng replied:
The current style is to not add braces around nested `if`/`for` statements if the innermost block is one line long. I chose to adhere to that style here, since the lines are short. We could discuss changing the style later.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -92,9 +92,36 @@ static bool isArrayLiteralIntrinsic(ApplyInst *ai) {
+  if (!ptai)
+    return nullptr;
+  // Return the `array.uninitialized_intrinsic` application, if it exists.
+  if (auto *dti = dyn_cast<DestructureTupleInst>(
```
Also braces here.

dan-zheng replied:
Done.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -92,9 +92,36 @@ static bool isArrayLiteralIntrinsic(ApplyInst *ai) {
+  return nullptr;
+}
+
+/// Given an element address from an `array.uninitialized_intrinsic` `apply`
```
Do literals with tuples get handled properly? For example, `[(x, y)]` gets initialized by extracting the addresses of the tuple elements and then storing `x` and `y` to them.

To handle that, this might need to understand tuple_element_addr instructions.

I think a similar thing might also happen for arrays of structs, though I couldn't trigger it with simple examples, so maybe it doesn't. Seems like it might be safer to handle all address extraction instructions anyways.

rxwei replied:
Great observation.  I don't think this will occur in reality though, because tuples cannot conform to protocols. Even if you make a custom type be `ExpressibleByArrayLiteral` where `ArrayLiteralElement == (Float, Float)`, there still doesn't seem to be any code path that would trigger it, because `init(arrayLiteralElements:)` cannot be marked `@differentiable`.

marcrasi replied:
Oh, seems like the tuple case is safeish! Maybe there is some struct case that will break it? I tried a few simple examples, and I couldn't make an array literal initialization that uses struct_element_addr, but I feel like it might happen sometimes.

marcrasi reviewed (commented):
I haven't finished reading this but here is an initial thought that I had.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1644,12 +1671,19 @@ bool LinearMapInfo::shouldDifferentiateInstruction(SILInstruction *inst) {
-#define CHECK_INST_TYPE_ACTIVE_DEST(INST) \
-  if (auto *castInst = dyn_cast<INST##Inst>(inst)) \
-    return activityInfo.isActive(castInst->getDest(), indices);
+  // For `store`-like instructions whose destination is an element address from
```
(Can be revisited later if you'd rather submit this improvement now instead of spending more time.)

Alternative to this new activity logic that might be less code and more robust: Modify variedness propagation to propagate variedness from the address to the array, and modify usefulness propagation to propagate usefulness from the array to the address.

This will make it robust to the thing about tuples that I commented about above. (Which might not actually be a problem as noted by @rxwei, but it still seems like it would be good to propagate activity though these things in case there are situations we haven't noticed.)

This will also make the addresses and projections show up as active in the debugging output, which will make it easier to understand. (I assume that they don't show up as active now.)

Do you think this approach is good and will work?

dan-zheng replied:
Thanks for the suggestion!

Element addresses (`pointer_to_address` and `index_addr`) of active `array.uninitialized_intrinsic` applications are now marked active in https://github.com/apple/swift/pull/28320/commits/9558aa2a7ed9cc3c83d7b5f4531f3bfbd482f6e8. This makes sense semantically and simplifies `LinearMapInfo::shouldDifferentiateInstruction`.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -6759,145 +6793,144 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
+  /// Given a `store` or `copy_addr` instruction whose destination is an element
+  /// address from an `array.uninitialized_intrinsic` application, accumulate
+  /// array element adjoint into the source's adjoint.
+  void accumulateArrayElementAdjoint(SILInstruction *inst) {
```
(Can be revisited later if you'd rather submit this improvement now instead of spending more time.)

Can this be combined with the logic for the normal stores/copies, by adding an array case to `getAdjointProjection`? If doable, this would slightly reduce the amount of code and would also take advantage of the recursiveness of `getAdjointProjection` to handle the tuple_element_addr/struct_element_addr case that I keep mentioning.

dan-zheng replied:
> Can this be combined with the logic for the normal stores/copies, by adding an array case to `getAdjointProjection`? If doable, this would slightly reduce the amount of code and would also take advantage of the recursiveness of `getAdjointProjection` to handle the tuple_element_addr/struct_element_addr case that I keep mentioning.

Good suggestion! I filed TF-976 to track this improvement.

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -6759,145 +6793,144 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
     // When the store's source is a `copy_value`, the `copy_value` is part of
     // array literal initialization. In this case, add the adjoint to the source
     // of the copy directly.
     if (auto *cvi = dyn_cast<CopyValueInst>(src))
```
Is this necessary for correctness, or is it just an optimization? Could you add this info to the comment?

dan-zheng replied:
The special `copy_value` source of `store` logic was added in https://github.com/apple/swift/pull/26737

The PR description suggests that the logic is necessary for correctness, but all tests continue to pass if the code is removed. I'll remove the code.

marcrasi replied:
Sounds good!

Based on a quick glance at that PR, it looks like that used to be necessary for correctness because of a similar problem to the problem that the current PR is solving. So now that you have solved the problem more generally, this special case is no longer necessary. Yay for simplification.

marcrasi reviewed (approved):
Yay!

dan-zheng committed:
Mark `array.uninitialized_intrinsic` application element addresses as useful.

This makes sense semantically and simplifies
`LinearMapInfo::shouldDifferentiateInstruction`.

dan-zheng committed:
Delete old special case in `PullbackEmitter::accumulateArrayElementAdjointDirect`.

Checking for `copy_value` source of `store` instruction no longer seems necessary.

dan-zheng committed:
Gardening.

dan-zheng committed:
Add tests.

Add nested array literal tests.
Add diagnostic test for array literal with tuple element.
Add negative tests:
- TF-975: multiple array literals.
- TF-977: array literals with `tuple_element_addr` elements.
- TF-978: apply with array literal address as indirect result.

marcrasi commented:
I thought of another case that might not be handled yet: creating array literals from indirect results. Here's a test for it.

```
struct GenericContainer<T: Differentiable>: Differentiable {
  var v: T
}

@differentiable
func arrayLiteralFromIndirectResult<T>(_ x: T, _ y: T) -> [GenericContainer<T>] {
  return [GenericContainer(v: x), GenericContainer(v: y)]
}

let pb = pullback(at: Float(0), Float(0)) {
  arrayLiteralFromIndirectResult($0, $1)
}
let tv = pb([GenericContainer<Float>.TangentVector(v: 1), GenericContainer<Float>.TangentVector(v: 2)])

// Should be (1.0, 2.0).
print(tv)
```

rxwei commented:
Sigh, I guess this would also require adding this special case for `apply` instructions whose indirect results are an array literal buffer.

dan-zheng commented:
> I thought of another case that might not be handled yet: creating array literals from indirect results. Here's a test for it.

Thanks! The test currently fails, I'll add a negative test.

> Sigh, I guess this would also require adding this special case for `apply` instructions whose indirect results are an array literal buffer.

I suspect this will be fixed by applying @marcrasi's suggestion in https://github.com/apple/swift/pull/28320#discussion_r347574989 to move "get adjoint buffer of an array literal element" logic out of `accumulateArrayElementAdjoint` into `getAdjointProjection`. `getAdjointBuffer` on the `apply` instruction's indirect result should get the array literal element adjoint buffer. TF-976 tracks this.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
Merging to unblock progress. Known issues and improvements are listed in the PR description.
Thanks reviewers!
